### PR TITLE
Better Error Reporting & a separate Error Log

### PR DESCRIPTION
**Pull request by Sharma-Hrishabh**
PR addressing #3388

ToDo:
- [x] Handle Hyperlinks in Unsaved Files.
- [x] Set the Default location of ErrorLog Widget as tabbed.

Work Done:

Step-1:
- [x] Make a new function like PRINT, and pass different arguments individually, or we can use JSON objects and pass it after stringifying.
- [x] A separate class for ERRORS and warnings will receive all the new messages.

Step-2: 
- [x] Make the console output compatible with the new print function.

Step-3:
- [x] Code Refactoring and remove the old version of PRINT from the codebase

Step-4:
- [x] Handle failing tests

Step-5:
- [x] Make the messages in log clickable to jump to the location.

Step-6:
- [x] Add a filter on the basis message groups in GUI

Step-7:
- [x] Hyperlink, console messages with their location.

### CODE REVIEW

File: src/ErrorLog.h [hunk 1]
```diff
@@ -11,10 +11,16 @@ class ErrorLog : public QWidget, public Ui::errorLogWidget
 
 public:
 	ErrorLog(QWidget *parent = nullptr);
+	ErrorLog(const ErrorLog& source) = delete;
+    ErrorLog(ErrorLog&& source) = delete;
```
Comment (tombsar) [inline src/ErrorLog.h:15]: Indentation type is wrong here.

Comment (Sharma-Hrishabh) [reply src/ErrorLog.h:15]: Thanks for pointing. I will fix it :)

File: src/mainwin.cc [hunk 2]
```diff
@@ -3047,9 +3097,21 @@ void MainWindow::consoleOutput(const QString &msg)
...
+	this->errorLogWidget->toErrorLog(log_msg);
+}
+
 void MainWindow::setCurrentOutput()
 {
 	set_output_handler(&MainWindow::consoleOutput, this);
+	set_output_handler2(&MainWindow::errorLogOutput,this);
 }
 
 void MainWindow::hideCurrentOutput()
```
Comment (t-paul) [inline src/mainwin.cc:3117]: Both `hideCurrentOutput()` and clearCurrentOutput()` need to do something with the 2nd handler too.

File: src/printutils.cc [hunk 3]
```diff
@@ -31,6 +37,11 @@ void set_output_handler(OutputHandlerFunc *newhandler, void *userdata)
 	outputhandler_data = userdata;
 }
 
+void set_output_handler2(OutputHandlerFunc2 *newhandler, void *userdata)
```
Comment (t-paul) [inline src/printutils.cc:40]: I wonder if it's better to combine the 2 setters, we don't expect different `userdata` (different MainWindow) in any case I think. So one pointer should be enough.

**Comment by Sharma-Hrishabh** [comment]
The message format for overwriting:
<a was assigned on line 5 but was overwritten> (in file test.scad, line 6)

<from args>(from loc)
 
So, it looks like a workaround, with what are you trying to say @t-paul 
Right?

File: src/printutils.cc (hunk 3 above)
Comment (Sharma-Hrishabh) [reply src/printutils.cc:41]: Done :)

File: src/mainwin.cc (hunk 2 above)
Comment (Sharma-Hrishabh) [reply src/mainwin.cc:3118]: Done

**Comment by rcolyer** [comment]
- The current hyperlink system for the error log is failing (and ungracefully) if used on an unsaved document.  It opens a new tab to the current directory from which OpenSCAD was launched, and then produces a "Failed to open" "file to open is a directory" error.
- The error log is showing up as split-window for the console, rather than as a tab, which is resulting in not enough horizontal space for either when the editor is open on the left of them.  Most IDEs use a tabbed system for this type of thing to provide enough display space.  In our case, upon preview or render the tab should then auto-switch to the error-log if an error or warning has occurred to draw attention to the issue that has likely scrolled past.

**Review by t-paul — APPROVED** [review]

**Comment by t-paul** [comment]
Imported to branch `error-info-window` in  #3435.

**Pull request merged**